### ISSUE TITLE
[question] how do i get the path to the exported or installed package in cache?

### ISSUE THREAD

**Issue by gsemet** [conan-io/conan#8871]
Hi,

I would like to know the path in the cache to the `package` folder of a given package name and id.

For example:
```
/home/someuser/.conan/data/my-package-name/0.2.1/user/channel/package/e106f0e3de7fc8e9a057009394741ef4441857ec/
```
I tried with reconstructing the path but on windows, or with short/long path, it makes it hard to rebuild.

I would like a command or a python function to call that would take the package full name and the package id and it returns me the path to the `package` folder in cache.

It seems like ConanApi.get_path does the job but I don't know how to call it from my script?

`conan info` does not return this information for a given package id, or I cannot get it work:
```
conan info my-package-name/0.2.1@user/channel:e106f0e3de7fc8e9a057009394741ef4441857ec
```
it does not work with package id

Gaetan

**Comment by gsemet** [comment]
Look similar to https://github.com/conan-io/conan/issues/6885

**Comment by solvingj** [comment]
> Look similar to #6885

I think that ticket is substantially different in nature. 


> it does not work with package id

simply put, no it does not. you must provide the profile information (settings/options/etc).

**Comment by gsemet** [comment]
Ok, how can I solve my problem ? I would like to find this path in Python I can call the Conan package directly, but how can I call the get_path function? How can I instantiate a conanapi object from my script (today I call the Conan executable, but instantiating the main Conan code from my code is possible but it looks too hacky for me)?

is this ger_path function that does what I need available on a more generic way?

**Comment by gsemet** [comment]
My use case is I need to binary compare the content of two package (the files in package without recipe, like headers and generated files). I did not find any reproducible hash I can use (I need if content is the same, hash is the same). I have deterministic binaries and generated files in my package. So comparison work fine, but on some weird windows environment finding the package path in the cache is different that under Linux, so I am looking for a generic solution. Once I get the path in cache, everything else is done.

**Comment by gsemet** [comment]
I fixed my use case by implementing the following function in my Conan wrapper :
```python

import pathlib

from conans.client.conan_api import ConanAPIV1
from conans.client.tools.win import unix_path
from conans.model.ref import ConanFileReference
from conans.model.ref import PackageReference

def get_package_folder_in_cache(package_name: str, package_id: str) -> pathlib.Path:
    conan_api = ConanAPIV1()
    conan_api.create_app(quiet_output=False)
    ref = ConanFileReference.loads(package_name)
    package_layout = conan_api.app.cache.package_layout(ref, short_paths=None)

    pref = PackageReference(
        package_layout._ref, package_id  # pylint: disable=protected-access
    )
    pkg_folder = package_layout.package(  # pylint: disable=assignment-from-no-return
        pref
    )
    return pathlib.Path(unix_path(pkg_folder))
```

I know it is not elegant and may break at any time, but this is the only way I found to fetch the real package folder using the same layout mechanism that what conan actually use.

**Comment by MihaiPopescu1985** [comment]
Try this:
conan info --paths --package-filter <package_name> --only "export_folder"  .

**Comment by MihaiPopescu1985** [comment]
Try this:
conan info --paths --package-filter <package name> --only "package_folder"  <conan recipe>

**Comment by MihaiPopescu1985** [comment]
conan info --paths --package-filter my_package/1.0.0@mine/org --only "package_folder"  C:\path\to\my\recipe\that\uses\this\package\conanfile.py

**Comment by retanoj** [comment]
> conan info --paths --package-filter my_package/1.0.0@mine/org --only "package_folder" C:\path\to\my\recipe\that\uses\this\package\conanfile.py

@MihaiPopescu1985  @memsharded 

so how could I do this in conan 2 ?

**Comment by memsharded** [comment]
Conan 2.0 provides a new ``conan cache path <ref>`` command, which would make obtaining paths very straightforward: https://docs.conan.io/2/reference/commands/cache.html


I think this command solves the OP perfectly, so closing this ticket as resolved. Please create new tickets if you have any further questions.

**Comment by simon-rechermann** [comment]
@memsharded 
I would like to get the package_folder path in the conan cache with conan 2 by providing a conanfile.

`conan cache path <ref>` does not work with a <path_to_conanfile> but just with a \<reference\>

In conan 1 I could do the following:
`conan info <path_to_conanfile> --paths --only package_folder --package-filter <package_name>*`

So I didn't have to specify the version@user/channel in the command but just the <package_name>.
Is it possible to extend the `conan cache path` command or is there another possibility get the conan cache path to a package by specifying the <path_to_conanfile> in conan 2?

I tried `conan graph info <path_to_conanfile> --package-filter <package_name>* --filter package_folder` but the package_folder field is None. Only the recipe_folder is set.

**Comment by memsharded** [comment]
Hi @simon-rechermann 

> conan cache path <ref> does not work with a <path_to_conanfile> but just with a <reference>

A ``path-to-conanfile`` is not a package in the cache, it is impossible to get the path to the package if it is not a package. But if I understood correctly you want to get the package-folder of the dependencies of the current ``path_to_conanfile.py``?

> tried conan graph info <path_to_conanfile> --package-filter <package_name>* --filter package_folder but the package_folder field is None. Only the recipe_folder is set.

Yes, this is expected. The ``conan graph info`` doesn't install (download or build) binaries at all, so it is not possible to have package folders for them. If you instead "install", then in Conan 2 you can:

```
conan install . --format=json > graph.json
```
The resulting ``graph.json`` will contain all the information of the graph, including the "package_folder" of all dependencies.

If you have any further question, it might be better to open a new ticket instead, closed tickets are more difficult to track. Thanks!
